udsoncan/services/Authentication.py

```diff
@@ -0,0 +1,383 @@
+            data = Authentication._append_byes_parameter(data, proof_of_ownership_client, 'Proof Of Ownership Client')
+            # challengeClient
+            data = Authentication._append_byes_parameter(data, challenge_client, 'Challenge Client')
+            # additionalParameter
+            data = Authentication._append_byes_parameter(data, additional_parameter, 'Additional Parameter')
+
+        return Request(service=cls, subfunction=authentication_task, data=data)
+
+    @staticmethod
+    def _extract_byes_parameter(response: Response, data: bytes, parameter_name: str):
+        if data is not None and len(data) >= 2:
+            parameter_length, = struct.unpack('>H', data[:2])
+        else:
+            raise InvalidResponseException(response, f'Not enough data for length of {parameter_name}')
+
+        if len(data) >= (2 + parameter_length):
+            return data[2:(2 + parameter_length)]
+        else:
+            raise InvalidResponseException(response, f'Not enough data for {parameter_name}.'
+                                                     f' Parsed length is: {parameter_length}, reset of payload length'
+                                                     f' is {len(data) - 2}')
+
+    @classmethod
+    def interpret_response(cls, response: Response) -> InterpretedResponse:
+        """
+        Populates the response ``service_data`` property with an instance of
+         :class:`Authentication.ResponseData<udsoncan.services.Authentication.ResponseData>`
+
+        :param response: The received response to interpret
+        :type response: :ref:`Response<Response>`
+
+        :raises InvalidResponseException: If response data length is not as expected
+        """
+
+        if response.data is None:
+            raise InvalidResponseException(response, "No data in response")
+
+        if len(response.data) < 2:
+            raise InvalidResponseException(response, "Response data must be at least 2 bytes")
+
+        subfunction = response.data[0]
+        return_value = response.data[1]
+        offset = 2
+        challenge_server: Optional[bytes] = None
+        ephemeral_public_key_server: Optional[bytes] = None
+        certificate_server: Optional[bytes] = None
+        proof_of_ownership_server: Optional[bytes] = None
+        session_key_info: Optional[bytes] = None
+        algorithm_indicator: Optional[bytes] = None
+        needed_additional_parameter: Optional[bytes] = None
+
+        if subfunction in (Authentication.AuthenticationTask.deAuthenticate,
+                           Authentication.AuthenticationTask.transmitCertificate,
+                           Authentication.AuthenticationTask.authenticationConfiguration):
+            pass  # no more parameters
+
+        elif subfunction == Authentication.AuthenticationTask.verifyCertificateUnidirectional:
+            challenge_server = Authentication._extract_byes_parameter(response, response.data[offset:],
```

Comment:
I know that this piece of code is good. But mypy is nto good enough to understand it. I try to make mypy happy as much as possible.

Can you add an `assert challenge_server is not None` before calling `len()`?.  This issue happens 12 times

```
udsoncan/services/Authentication.py:309: error: Argument 1 to "len" has incompatible type "bytes | None"; expected "Sized"  [arg-type]
udsoncan/services/Authentication.py:313: error: Argument 1 to "len" has incompatible type "bytes | None"; expected "Sized"  [arg-type]
udsoncan/services/Authentication.py:318: error: Argument 1 to "len" has incompatible type "bytes | None"; expected "Sized"  [arg-type]
udsoncan/services/Authentication.py:322: error: Argument 1 to "len" has incompatible type "bytes | None"; expected "Sized"  [arg-type]
udsoncan/services/Authentication.py:326: error: Argument 1 to "len" has incompatible type "bytes | None"; expected "Sized"  [arg-type]
udsoncan/services/Authentication.py:330: error: Argument 1 to "len" has incompatible type "bytes | None"; expected "Sized"  [arg-type]
udsoncan/services/Authentication.py:335: error: Argument 1 to "len" has incompatible type "bytes | None"; expected "Sized"  [arg-type]
udsoncan/services/Authentication.py:350: error: Argument 1 to "len" has incompatible type "bytes | None"; expected "Sized"  [arg-type]
udsoncan/services/Authentication.py:354: error: Argument 1 to "len" has incompatible type "bytes | None"; expected "Sized"  [arg-type]
udsoncan/services/Authentication.py:359: error: Argument 1 to "len" has incompatible type "bytes | None"; expected "Sized"  [arg-type]
udsoncan/services/Authentication.py:364: error: Argument 1 to "len" has incompatible type "bytes | None"; expected "Sized"  [arg-type]
```

Reply:
You can add a  `#type: ignore` next to the `len()` function, i'm good with that as well. mypy is wrong here.

Reply:
the function just missing annotation regarding the return value type, will fix

Reply:
ooohhh, right! good catch